src/v/datalake/catalog_schema_manager.cc

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/catalog_schema_manager.h"
+
+#include "base/vlog.h"
+#include "datalake/logger.h"
+#include "datalake/table_definition.h"
+#include "iceberg/field_collecting_visitor.h"
+#include "iceberg/table_identifier.h"
+#include "iceberg/transaction.h"
+
+namespace datalake {
+
+namespace {
+catalog_schema_manager::errc log_and_convert_catalog_err(
+  iceberg::catalog::errc e, std::string_view log_msg) {
+    switch (e) {
+    case iceberg::catalog::errc::shutting_down:
+        vlog(datalake_log.debug, "{}: {}", log_msg, e);
+        return catalog_schema_manager::errc::shutting_down;
+    case iceberg::catalog::errc::timedout:
+    case iceberg::catalog::errc::not_found:
+    case iceberg::catalog::errc::io_error:
+    case iceberg::catalog::errc::unexpected_state:
+    case iceberg::catalog::errc::already_exists:
+        vlog(datalake_log.warn, "{}: {}", log_msg, e);
+        return catalog_schema_manager::errc::failed;
+    }
+}
+enum class fill_errc {
+    // There is a mismatch in a field's type, name, or required.
+    mismatch,
+    // We couldn't fill all the columns, but the ones we could all matched.
+    incomplete,
+};
+// Performs a simultaneous, depth-first iteration through fields of the two
+// schemas, filling dest's field IDs with those from the source. Returns
+// successfully if all the field IDs in the destination type are filled.
+checked<std::nullopt_t, fill_errc>
+fill_field_ids(iceberg::struct_type& dest, const iceberg::struct_type& source) {
+    using namespace iceberg;
+    chunked_vector<nested_field*> dest_stack;
+    dest_stack.reserve(dest.fields.size());
+    for (auto& f : std::ranges::reverse_view(dest.fields)) {
+        dest_stack.emplace_back(f.get());
+    }
+    chunked_vector<nested_field*> source_stack;
+    source_stack.reserve(source.fields.size());
+    for (auto& f : std::ranges::reverse_view(source.fields)) {
+        source_stack.emplace_back(f.get());
+    }
+    while (!source_stack.empty()) {
+        if (dest_stack.empty()) {
```

Comment:
Can this be folded into the while loop condition?

Reply:
Good spot, done